src/persist-client/src/write.rs

```diff
@@ -966,4 +969,57 @@ mod tests {
         );
         assert_eq!(write.upper(), &Antichain::from_elem(7));
     }
+
+    #[mz_ore::test(tokio::test(flavor = "multi_thread"))]
+    #[cfg_attr(miri, ignore)] // unsupported operation: returning ready events from epoll_wait is not yet implemented
+    async fn fetch_recent_upper_linearized() {
+        type Timestamp = u64;
+        let max_upper = 1000;
+
+        let shard_id = ShardId::new();
+        let mut clients = PersistClientCache::new_no_metrics();
+        let upper_writer_client = clients.open(PersistLocation::new_in_mem()).await.unwrap();
+        let (mut upper_writer, _) = upper_writer_client
+            .expect_open::<(), (), Timestamp, i64>(shard_id)
+            .await;
+        // Clear the state cache between each client to maximally disconnect
+        // them from each other.
+        clients.clear_state_cache();
+        let upper_reader_client = clients.open(PersistLocation::new_in_mem()).await.unwrap();
+        let (mut upper_reader, _) = upper_reader_client
+            .expect_open::<(), (), Timestamp, i64>(shard_id)
+            .await;
+        let (tx, rx) = mpsc::channel();
+
+        let task = task::spawn(|| "upper-reader", async move {
+            let mut upper = Timestamp::MIN;
+
+            while upper < max_upper {
+                while let Ok(new_upper) = rx.try_recv() {
```

Comment:
@danhhz FYI, I update `if let ...` to `while let ...` after you approved. I think it's a bit more likely to catch a bug if it always is checking against the most recently known upper in a scenario where the channel gets backed up.

Reply:
Nice